In [30]:
import pandas as pd
#import matplotlib.pyplot as plt
import seaborn as sns

In [31]:
df = pd.read_csv('example.csv')
df

Unnamed: 0  Safety    yacEPA  Drive  FieldGoalDistance    airEPA  \
0          244485       0  1.240299     18                NaN -0.588101   
1          115340       0       NaN     22                NaN       NaN   
2           68357       0       NaN      8                NaN       NaN   
3          368377       0  1.075660     24                NaN  0.402080   
4          384684       0       NaN     11                NaN       NaN   
...           ...     ...       ...    ...                ...       ...   
40764       79438       0       NaN      3                NaN       NaN   
40765      398233       0       NaN      9                NaN       NaN   
40766      396466       0       NaN      8                NaN       NaN   
40767      333861       0  0.025040     14                NaN  0.280049   
40768      247608       0 -2.441755      6                NaN  1.692283   

       Field_Goal_Prob posteam  Penalty.Yards FieldGoalResult  ...  \
0             0.309404      TB              0             NaN  ...   
1             0.198142     OAK              0             NaN  ...   
2             0.000000     NYJ              0             NaN  ...   
3             0.258227     CLE              0             NaN  ...   
4             0.199771     DEN              0             NaN  ...   
...                ...     ...            ...             ...  ...   
40764         0.381506      NO              0             NaN  ...   
40765         0.244675     DEN              0             NaN  ...   
40766         0.147057     ATL              0             NaN  ...   
40767         0.373774     IND              0             NaN  ...   
40768         0.210732     CIN              0             NaN  ...   

       Accepted.Penalty qtr  AwayTimeouts_Remaining_Post  RunGap  Reception  \
0                     0   3                            3     NaN          1   
1                     0   4                            3     NaN          0   
2                     0   2                            1     NaN          0   
3                     0   4                            0     NaN          1   
4                     0   2                            2     end          0   
...                 ...  ..                          ...     ...        ...   
40764                 0   1                            3   guard          0   
40765                 0   2                            3   guard          0   
40766                 0   2                            1   guard          0   
40767                 0   3                            3     NaN          1   
40768                 0   1                            3     NaN          0   

       TimeUnder Season  Opp_Safety_Prob No_Score_Prob Challenge.Replay  
0              1   2014         0.000527      0.024134                0  
1              7   2011         0.001883      0.246929                0  
2              1   2010         0.000000      0.000000                0  
3              9   2017         0.000845      0.133814                0  
4             10   2017         0.003453      0.170787                0  
...          ...    ...              ...           ...              ...  
40764          3   2010         0.000021      0.005075                0  
40765          3   2017         0.000134      0.287528                0  
40766          1   2017         0.001132      0.612357                0  
40767          7   2016         0.000042      0.002641                0  
40768          2   2014         0.006138      0.040037                0  

[40769 rows x 32 columns]

In [32]:
df.describe()

Unnamed: 0        Safety        yacEPA         Drive  \
count   40769.000000  40769.000000  15867.000000  40769.000000   
mean   202916.414359      0.000294     -0.396137     12.287817   
std    117725.700085      0.017154      1.998663      7.169334   
min        47.000000      0.000000    -14.000000      1.000000   
25%    101844.000000      0.000000     -0.961252      6.000000   
50%    201785.000000      0.000000      0.000000     12.000000   
75%    304505.000000      0.000000      0.484938     18.000000   
max    407658.000000      1.000000      8.390143     34.000000   

       FieldGoalDistance        airEPA  Field_Goal_Prob  Penalty.Yards  \
count         902.000000  15875.000000     40748.000000   40769.000000   
mean           37.487805      0.522389         0.243680       0.607349   
std            10.049785      1.399720         0.159400       2.585225   
min            18.000000     -5.930319         0.000000       0.000000   
25%            29.000000     -0.503679         0.150847       0.000000   
50%            38.000000      0.295619         0.230632       0.000000   
75%            46.000000      1.369291         0.326247       0.000000   
max            66.000000      6.397495         0.994592      60.000000   

       DefTeamScore   Away_WP_pre  ...   RushAttempt  Accepted.Penalty  \
count  38019.000000  38223.000000  ...  40769.000000      40769.000000   
mean      11.366317      0.468696  ...      0.296745          0.072825   
std        9.901340      0.285774  ...      0.456829          0.259852   
min        0.000000      0.000000  ...      0.000000          0.000000   
25%        3.000000      0.235012  ...      0.000000          0.000000   
50%       10.000000      0.471362  ...      0.000000          0.000000   
75%       17.000000      0.679675  ...      1.000000          0.000000   
max       61.000000      1.000000  ...      1.000000          1.000000   

                qtr  AwayTimeouts_Remaining_Post     Reception     TimeUnder  \
count  40769.000000                 40769.000000  40769.000000  40769.000000   
mean       2.572371                     2.495867      0.248277      7.369717   
std        1.129827                     0.797636      0.432019      4.638820   
min        1.000000                    -1.000000      0.000000      0.000000   
25%        2.000000                     2.000000      0.000000      3.000000   
50%        3.000000                     3.000000      0.000000      7.000000   
75%        4.000000                     3.000000      0.000000     11.000000   
max        5.000000                     3.000000      1.000000     15.000000   

             Season  Opp_Safety_Prob  No_Score_Prob  Challenge.Replay  
count  40769.000000     40748.000000   40748.000000      40769.000000  
mean    2012.997547         0.002491       0.128504          0.008659  
std        2.577944         0.003425       0.200186          0.092649  
min     2009.000000         0.000000       0.000000          0.000000  
25%     2011.000000         0.000104       0.002724          0.000000  
50%     2013.000000         0.000977       0.024741          0.000000  
75%     2015.000000         0.003868       0.172581          0.000000  
max     2017.000000         0.028950       1.000000          1.000000  

[8 rows x 25 columns]

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40769 entries, 0 to 40768
Data columns (total 32 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Unnamed: 0                   40769 non-null  int64  
 1   Safety                       40769 non-null  int64  
 2   yacEPA                       15867 non-null  float64
 3   Drive                        40769 non-null  int64  
 4   FieldGoalDistance            902 non-null    float64
 5   airEPA                       15875 non-null  float64
 6   Field_Goal_Prob              40748 non-null  float64
 7   posteam                      38219 non-null  object 
 8   Penalty.Yards                40769 non-null  int64  
 9   FieldGoalResult              908 non-null    object 
 10  DefTeamScore                 38019 non-null  float64
 11  Tackler2                     4943 non-null   object 
 12  Away_WP_pre                  38223 non-null  float64
 13  Timeout_Indicato

In [59]:
df.drop(columns='TwoPointConv')

Unnamed: 0  Safety    yacEPA  Drive  FieldGoalDistance    airEPA  \
0          244485       0  1.240299     18                NaN -0.588101   
1          115340       0       NaN     22                NaN       NaN   
2           68357       0       NaN      8                NaN       NaN   
3          368377       0  1.075660     24                NaN  0.402080   
4          384684       0       NaN     11                NaN       NaN   
...           ...     ...       ...    ...                ...       ...   
40764       79438       0       NaN      3                NaN       NaN   
40765      398233       0       NaN      9                NaN       NaN   
40766      396466       0       NaN      8                NaN       NaN   
40767      333861       0  0.025040     14                NaN  0.280049   
40768      247608       0 -2.441755      6                NaN  1.692283   

       Field_Goal_Prob posteam  Penalty.Yards FieldGoalResult  ...  RunGap  \
0             0.309404      TB              0             NaN  ...     NaN   
1             0.198142     OAK              0             NaN  ...     NaN   
2             0.000000     NYJ              0             NaN  ...     NaN   
3             0.258227     CLE              0             NaN  ...     NaN   
4             0.199771     DEN              0             NaN  ...     end   
...                ...     ...            ...             ...  ...     ...   
40764         0.381506      NO              0             NaN  ...   guard   
40765         0.244675     DEN              0             NaN  ...   guard   
40766         0.147057     ATL              0             NaN  ...   guard   
40767         0.373774     IND              0             NaN  ...     NaN   
40768         0.210732     CIN              0             NaN  ...     NaN   

      Reception  TimeUnder  Season  Opp_Safety_Prob  No_Score_Prob  \
0             1          1    2014         0.000527       0.024134   
1             0          7    2011         0.001883       0.246929   
2             0          1    2010         0.000000       0.000000   
3             1          9    2017         0.000845       0.133814   
4             0         10    2017         0.003453       0.170787   
...         ...        ...     ...              ...            ...   
40764         0          3    2010         0.000021       0.005075   
40765         0          3    2017         0.000134       0.287528   
40766         0          1    2017         0.001132       0.612357   
40767         1          7    2016         0.000042       0.002641   
40768         0          2    2014         0.006138       0.040037   

      Challenge.Replay  posteam_no_NaN new_drive  new_Field_Goal_Prob  
0                    0              TB        12              0.24368  
1                    0             OAK        12              0.24368  
2                    0             NYJ        12              0.24368  
3                    0             CLE        12              0.24368  
4                    0             DEN        12              0.24368  
...                ...             ...       ...                  ...  
40764                0              NO        12              0.24368  
40765                0             DEN        12              0.24368  
40766                0             ATL        12              0.24368  
40767                0             IND        12              0.24368  
40768                0             CIN        12              0.24368  

[40769 rows x 34 columns]

#### Two pointy Conv ustunida 40769 ta qatordan faqat 53 tasida ma'lumot bo'lib qolganlari Null bo'lgani uchun o'sha ustun tushirib yuborildi


In [34]:
print(df.isnull().sum())

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                         24073
Touchdown                          0
TwoPointConv                   40716
PassOutcome                    24018
PlayTimeDiff                      43
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

In [35]:
df.shape

(40769, 32)

In [36]:
df.dropna(how='any').shape

(0, 32)

#### Demak jadvalimizda umuman NaN yoki Null qatnashmagan qator mavjud emas

In [37]:
df.groupby('posteam').count()

Unnamed: 0  Safety  yacEPA  Drive  FieldGoalDistance  airEPA  \
posteam                                                                 
ARI            1208    1208     530   1208                 26     530   
ATL            1167    1167     492   1167                 33     493   
BAL            1218    1218     510   1218                 37     511   
BUF            1186    1186     477   1186                 29     476   
CAR            1128    1128     443   1128                 19     446   
CHI            1188    1188     465   1188                 18     465   
CIN            1189    1189     482   1189                 32     483   
CLE            1117    1117     480   1117                 21     480   
DAL            1192    1192     483   1192                 37     484   
DEN            1284    1284     544   1284                 33     544   
DET            1243    1243     590   1243                 32     591   
GB             1198    1198     524   1198                 26     524   
HOU            1209    1209     514   1209                 29     514   
IND            1218    1218     553   1218                 21     553   
JAC             936     936     392    936                 18     392   
JAX             272     272     122    272                  7     122   
KC             1210    1210     440   1210                 24     441   
LA              255     255     100    255                  6     100   
LAC             115     115      46    115                  5      46   
MIA            1160    1160     474   1160                 28     473   
MIN            1182    1182     453   1182                 26     453   
NE             1216    1216     517   1216                 38     517   
NO             1233    1233     548   1233                 30     548   
NYG            1181    1181     517   1181                 18     517   
NYJ            1183    1183     470   1183                 27     470   
OAK            1180    1180     478   1180                 24     478   
PHI            1244    1244     501   1244                 28     502   
PIT            1147    1147     481   1147                 35     481   
SD             1086    1086     446   1086                 30     446   
SEA            1211    1211     474   1211                 21     474   
SF             1207    1207     447   1207                 41     447   
STL             968     968     402    968                 17     402   
TB             1149    1149     495   1149                 36     495   
TEN            1138    1138     496   1138                 23     496   
WAS            1201    1201     481   1201                 27     481   

         Field_Goal_Prob  Penalty.Yards  FieldGoalResult  DefTeamScore  ...  \
posteam                                                                 ...   
ARI                 1206           1208               27          1202  ...   
ATL                 1167           1167               35          1163  ...   
BAL                 1217           1218               37          1211  ...   
BUF                 1186           1186               29          1181  ...   
CAR                 1128           1128               19          1121  ...   
CHI                 1188           1188               18          1180  ...   
CIN                 1188           1189               32          1184  ...   
CLE                 1116           1117               21          1110  ...   
DAL                 1192           1192               37          1185  ...   
DEN                 1284           1284               33          1278  ...   
DET                 1243           1243               32          1238  ...   
GB                  1195           1198               26          1191  ...   
HOU                 1209           1209               29          1203  ...   
IND                 1218           1218               21          1214  ...   
JAC                  936            936    

In [38]:
df['posteam_no_NaN'] = df['posteam'].fillna(value="None given")
df

Unnamed: 0  Safety    yacEPA  Drive  FieldGoalDistance    airEPA  \
0          244485       0  1.240299     18                NaN -0.588101   
1          115340       0       NaN     22                NaN       NaN   
2           68357       0       NaN      8                NaN       NaN   
3          368377       0  1.075660     24                NaN  0.402080   
4          384684       0       NaN     11                NaN       NaN   
...           ...     ...       ...    ...                ...       ...   
40764       79438       0       NaN      3                NaN       NaN   
40765      398233       0       NaN      9                NaN       NaN   
40766      396466       0       NaN      8                NaN       NaN   
40767      333861       0  0.025040     14                NaN  0.280049   
40768      247608       0 -2.441755      6                NaN  1.692283   

       Field_Goal_Prob posteam  Penalty.Yards FieldGoalResult  ...  qtr  \
0             0.309404      TB              0             NaN  ...    3   
1             0.198142     OAK              0             NaN  ...    4   
2             0.000000     NYJ              0             NaN  ...    2   
3             0.258227     CLE              0             NaN  ...    4   
4             0.199771     DEN              0             NaN  ...    2   
...                ...     ...            ...             ...  ...  ...   
40764         0.381506      NO              0             NaN  ...    1   
40765         0.244675     DEN              0             NaN  ...    2   
40766         0.147057     ATL              0             NaN  ...    2   
40767         0.373774     IND              0             NaN  ...    3   
40768         0.210732     CIN              0             NaN  ...    1   

      AwayTimeouts_Remaining_Post  RunGap  Reception  TimeUnder  Season  \
0                               3     NaN          1          1    2014   
1                               3     NaN          0          7    2011   
2                               1     NaN          0          1    2010   
3                               0     NaN          1          9    2017   
4                               2     end          0         10    2017   
...                           ...     ...        ...        ...     ...   
40764                           3   guard          0          3    2010   
40765                           3   guard          0          3    2017   
40766                           1   guard          0          1    2017   
40767                           3     NaN          1          7    2016   
40768                           3     NaN          0          2    2014   

      Opp_Safety_Prob  No_Score_Prob Challenge.Replay posteam_no_NaN  
0            0.000527       0.024134                0             TB  
1            0.001883       0.246929                0            OAK  
2            0.000000       0.000000                0            NYJ  
3            0.000845       0.133814                0            CLE  
4            0.003453       0.170787                0            DEN  
...               ...            ...              ...            ...  
40764        0.000021       0.005075                0             NO  
40765        0.000134       0.287528                0            DEN  
40766        0.001132       0.612357                0            ATL  
40767        0.000042       0.002641                0            IND  
40768        0.006138       0.040037                0            CIN  

[40769 rows x 33 columns]

#### posteam ustunidagi NaN larni o'rni ustun type i string bo'lgani ucun None given bilan to'ldirib chiqildi

In [39]:
key = df['yacEPA']
print(key.mean())
print(key.median())

-0.39613653499874507
0.0


In [45]:
key = df['Field_Goal_Prob']
print(key.mean())
print(key.median())

0.24367961032922636
0.2306319714387671


In [52]:
df['new_Field_Goal_Prob'] = key.fillna(key).mean()


In [53]:
print(df['new_Field_Goal_Prob'].mean())
print(df['new_Field_Goal_Prob'].median())

0.24367961032922641
0.24367961032922636


#### Field_Goal_Prob ustunidagi Nan lar o'rniga o'rtachasi qo'yilgan holda 'new_Field_Goal_Prob' ustuniga ko'chirildi. mean va medianlar deyarli o'zgarmagan

In [60]:
key = df['DefTeamScore']
print(key.mean())
print(key.median())

11.366316841579211
10.0


In [61]:
df['new_DefTeamScore'] = key.fillna(key).mean()


In [62]:
print(df['new_DefTeamScore'].mean())
print(df['new_DefTeamScore'].median())

11.366316841579211
11.366316841579211


#### Deyarli mean max o'zgarmadi
